ext/gmp/gmp.c

```diff
@@ -1293,14 +1302,16 @@ ZEND_FUNCTION(gmp_fact)
 		FETCH_GMP_ZVAL(gmpnum, a_arg, temp_a, 1);
 		FREE_GMP_TEMP(temp_a);
 
-		if (mpz_sgn(gmpnum) < 0) {
-			zend_argument_value_error(1, "must be greater than or equal to 0");
+		(void)gmpnum;
+		val = zval_get_long(a_arg);
```

Comment:
I don't quite understand how `zval_get_long()` fits into the mix here. At least, its usage makes `FETCH_GMP__ZVAL()` above superflous, and we wouldn't even need the special casing for `Z_TYPE_P(a_arg) == IS_LONG` (since this is already catered to by `zval_get_long()`.

What happens if one calls `gmp_fact("18446744073709551617")`? Wouldn't that raise a deprecation notice ('Implicit conversion from float-string "18446744073709551617" to int loses precision') instead of throwing an `ValueError`.

Reply:
> What happens if one calls `gmp_fact("18446744073709551617")`? Wouldn't that raise a deprecation notice ('Implicit conversion from float-string "18446744073709551617" to int loses precision') instead of throwing an `ValueError`.

Nope I get the exception in this case too. I ll see the rest later.

Reply:
Ah, indeed! That is because the conversion happens in coercive mode. And that also shows that we can't easily drop the seemingly superfluous `FETCH_GMP_ZVAL()` since only that will report issues regarding strict typing mode (e.g. passing a float). So `zval_get_long()` is basically correct here; the saturating behavior is fine since we cannot compute the factorial of `PHP_INT_MAX` anyway.

<details><summary>Still, there is an issue unrelated to this PR</summary>

E.g. for `gmp_fact(new GMP("18446744073709551617"))` `zval_get_long()` returns 1 which passes the overflow check, and evaluates to `GMP(1)`. That is because the `mpz_get_si()` in `gmp_cast_object()` overflows, while I would expect saturating behavior. This is not related to this PR, though, but rather a general issue with ext/gmp. From the [`mpz_get_si()` documentation](https://gmplib.org/manual/Converting-Integers):

> If op fits into a signed long int return the value of op. Otherwise return the least significant part of op, with the same sign as op.
>
> If op is too big to fit in a signed long int, the returned result is probably not very useful. To find out if the value will fit, use the function mpz_fits_slong_p.

So apparently, there is an easy fix for `gmp_cast_object()`. However, on LLP64 mpir 3.0.0 `mpz_fits_slong_p()` returns false for 2147483648. ;)

</details>

Reply:
> I don't quite understand how `zval_get_long()` fits into the mix here. At least, its usage makes `FETCH_GMP__ZVAL()` above superflous, and we wouldn't even need the special casing for `Z_TYPE_P(a_arg) == IS_LONG` (since this is already catered to by `zval_get_long()`.
> 
> What happens if one calls `gmp_fact("18446744073709551617")`? Wouldn't that raise a deprecation notice ('Implicit conversion from float-string "18446744073709551617" to int loses precision') instead of throwing an `ValueError`.

The `zval_get_long()` is to convert the GMP object back to an int...

Reply:
> The `zval_get_long()` is to convert the GMP object back to an int...

Yeah, I figured that much, but I wondered about values outside of `[ZEND_LONG_MIN, ZEND_LONG_MAX]`. However, I [found out](https://github.com/php/php-src/pull/16880#discussion_r1852240822). :)